src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -0,0 +1,259 @@
+            {
+                return _automaticRedirection;
+            }
+
+            set
+            {
+                CheckDisposedOrStarted();
+                _automaticRedirection = value;
+            }
+        }
+
+        #endregion
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing && !_disposed)
+            {
+                _disposed = true;
+            }
+
+            base.Dispose(disposing);
+        }
+
+        protected internal override Task<HttpResponseMessage> SendAsync(
+            HttpRequestMessage request,
+            CancellationToken cancellationToken)
+        {
+            if (request == null)
+            {
+                throw new ArgumentNullException("request", SR.net_http_handler_norequest);
+            }
+            if (request.RequestUri.Scheme == UriSchemeHttps)
+            {
+                throw NotImplemented.ByDesignWithMessage("HTTPS stack is not yet implemented");
+            }
+            if (request.Content != null)
+            {
+                throw NotImplemented.ByDesignWithMessage("HTTP requests with a body are not yet supported");
+            }
+
+            CheckDisposed();
+
+            SetOperationStarted();
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return Task.FromCanceled<HttpResponseMessage>(cancellationToken);
+            }
+
+            // Create RequestCompletionSource object and save current values of handler settings.
+            RequestCompletionSource state = new RequestCompletionSource();
+            state.CancellationToken = cancellationToken;
+            state.RequestMessage = request;
+            state.Handler = this;
+
+            Task.Factory.StartNew(
+                s =>
+                {
+                    var rcs = (RequestCompletionSource)s;
+                    rcs.Handler.StartRequest(rcs);
+                },
+                state,
+                CancellationToken.None,
+                TaskCreationOptions.DenyChildAttach,
+                TaskScheduler.Default);
+
+            return state.Task;
+        }
+
+        #region Private methods
+
+        private void StartRequest(RequestCompletionSource state)
+        {
+            HttpResponseMessage responseMessage = null;
+            SafeCurlHandle requestHandle = null;
+
+            if (state.CancellationToken.IsCancellationRequested)
+            {
+                state.TrySetCanceled();
+                return;
+            }
+
+            var cancellationTokenRegistration = state.CancellationToken.Register(x => ((RequestCompletionSource) x).TrySetCanceled(), state);
+
+            try
+            {
+                requestHandle = CreateRequestHandle(state);
+
+                state.CancellationToken.ThrowIfCancellationRequested();
+                int result = Interop.libcurl.curl_easy_perform(requestHandle);
+                if (CURLcode.CURLE_OK != result)
```

Comment:
There are three places in this try block that you throw an exception just to transfer flow to the catch block. Call HandleAsyncException directly from this if block and then return. Similarly for ThrowIfCancellationRequested, check IsCancellationRequested instead and call HandleAsycException directly.